### ISSUE TITLE
Build installer version

### ISSUE THREAD

**Issue by dnavarrocueto** [gama-platform/gama#1841]
It would be very interesting that Gama could build a .jar file of a model, like Repast does, to distribute it as an installable version to other people.
Thanks
Daniel

**Comment by AlexisDrogoul** [comment]
Agree. It's quite a huge effort, but it is something that we have to think of. Do not hesitate in adding more comments about the features you could want (like an editor or no editor, etc.), which "minimal" version of the views/displays/experiment environment you would like to be able to manipulate, how embedded a model should be (e.g. giving the possibility to change data files or not ?)

**Comment by dnavarrocueto** [comment]
I don't think an editor is needed. If someone wants to edit a model you work on the code with the Gama platform, which is very powerful. But some users just want to install and run the simulation changing only some parameters.

Also it would be fine to decide when to distribute the code. Sometimes you are working on an open environment so you share the code, but sometimes that is not the case.

So, I suggest that the interface of the .jar installable version of your model has only the 'simulation perspective' and not the 'modeling perspective'.

**Comment by AlexisDrogoul** [comment]
Note that this issue supersedes #742 (closed a long time ago, but never addressed).

**Comment by dnavarrocueto** [comment]
As a suggestion, GAMA may use IzPack.

**Comment by sriramab** [comment]
I think it should have a reload button that can take changed parameters from users (user-buttons, sliders, input boxes, dropdowns). For being able to change file inputs, I would say the developer can instruct the user to have certain files in the folder of this executable to be able to run it.  I feel console and interactive console are not necessary.

**Comment by sriramab** [comment]
May be as an intermediate step before the full standalone, it could be a (just example) a .gamr (gama run file) instead of .gaml (gama language file) . On double clicking, it is opened in GAMA.app but without consoles, without model exploer, without error window, without editor window. so it is just displays (2/3 screen) and parameter window (1/3 screen)

**Comment by NeKroZ** [comment]
Agreed, i think this IS one of the most lacking features this platform.

**Comment by agrignard** [comment]
Hi everyone,

We have been discussing this with Patrick and Benoit this week and we were wondering what was the status on that. Is there any update on it. Indeed it would be greta to have stand alone model where the user is only playing with the parameter.

How difficult is it to implement it?

**Comment by AlexisDrogoul** [comment]
Status is not addressed. And I guess that if @hqnghi88 continues to progress on the web version of GAMA, it will never be (as the desktop version will be something of the past :) ). 

More seriously, embedding GAMA into a model is in my view a nonsense (what would be the advantage of carrying 80Mb++ of binaries just to avoid downloading two different things ?). So, w.r.t. to what is possible now, @sriramab suggestion (having self-contained model "files" that are treated in a special way, automatically, by the desktop version of GAMA) is probably the best way to go. A .gamar file could for example be the zipped version of a project (with all the support files required: shape files, etc.) that, when launched, would be executed automatically instead of launching the editor, etc. 

Everything is more or less in place to make this possible. It just needs to be coded...

**Comment by hqnghi88** [comment]
Dear all,
An ideal world that GAMA could stay on the cloud, such world that i'm dreaming. But i think no big progress in near future. Anyway, i have headless comodeling and the Rjava for this weekend, so may be i'll return next week with gamar. As Alexis said, everything is in place , just need to be coded.

**Comment by sriramab** [comment]
👍  just today I was asked in a meeting on a new project, if "can we just play around the model without seeing any of the ****** code" ... 😋 . 

I hope some non-web solution also emerges from this. 

I just have one suggestion to add, there should be an option **not** to include (supplementary) shape files/data files into gamar. Sometimes shape files and data files themselves are so huge. Once you are able to send your gamar file with a good  project folder structure, the recipient should place the required (supplementary) files into the ../includes folder .. may be sent via email or dropbox or other services.

Thank you

**Comment by agrignard** [comment]
For different reason I think it's important to have it on a user point of view (on a developper point of view yes it's useless) and it is independant with the web browser version

0. For a non computer scientist, he might maybe not be able to understand the code and also maybe you don't want him to potentially change the code and introduce bug

1. If you have a user that just want to play, demo explore the model with the parameters expose

2. For a potential client you don't want to expose the code and you can then choose what you wan to expose to him (change shapefile, input through csv, etc) but you control it.

I know it's not the ideal solution but worse case it's not a big deals nowadays to download a 100Mb executable, I mean it's better if it's a like one but it's not a big deal.

Typically having this interface as an .exe would be great!
<img width="1319" alt="screen shot 2017-09-28 at 16 04 25" src="https://user-images.githubusercontent.com/3928502/30970928-c63c7480-a466-11e7-8b83-6045f30bcd3a.png">

**Comment by AlexisDrogoul** [comment]
Well. What Srirama proposes is more or less the same for end users.
Embedding the runtime is pointless. Who (except netlogo, maybe! ) embeds a
jdk with a jar file or an applet nowadays?

The other point is more philosophical and related to research. People who
insist on simply using a model without wanting to understand how it works
are in my opinion the kinds of users we should avoid or convince of the
necessity they have to make an effort. These are usually the same people
that believe models will provide them with 'solutions' or that the modeling
step is not important, only the 'simulation' one is.

We should in my opinion put more efforts in training, convincing, guiding
users so that they understand that designing and writing a model is
actually crucial in complex problems solving (and sometimes more useful for
finding solutions than running the model itself) , rather than removing
modules so that these users are not too much disrupted in their old habits.
This probably means, also that more intuitive ways of programming and
modeling should be proposed (if there is a lack in gama, I see it here)



Le 28 sept. 2017 21:12, "Arnaud Grignard" <notifications@github.com> a
écrit :

**Comment by hqnghi88** [comment]
![gamr](https://user-images.githubusercontent.com/6105384/31310414-ecb202f8-abc1-11e7-8d72-8a71147e2125.gif)
Hello gamr, i have a draft gamr, but as Alexis said we need a preinstalled gama, not a embedded in zip file, and i wonder how to generate a gamr, currently i zipped a gama project. Would it be a menu in Gama?

**Comment by AlexisDrogoul** [comment]
Well done Nghi. I wonder how it works ? How you specify the experiment to run ? If you use the `auto-run` facet, does it run automatically ? My initial idea was to use a `.experiment` file and let the modeler save the project as a `runtime` file (a zipped project as well), with only this experiment available for running.

**Comment by hqnghi88** [comment]
Dear Alexis,
Totally as same as your idea, but i use a xml in zip package (renamed to .gamr) to specify which gaml file and experiment, using your program param which launch Gama with file type gamr. THen , unzip to system temporary folder . I think use .experiment file is the best solution, but i did not find the structure of this file type and how can we create it?

**Comment by AlexisDrogoul** [comment]
Well, it's one of the two files types you can create in GAMA :)

In the navigator, select "New..." and then "Experiment".

**Comment by agrignard** [comment]
Wha it looks great!! How can we test this?

**Comment by agrignard** [comment]
I am thinking also that in such a feature it would be nice to have the option to "embed" the GAMA version that works for the specific model. So at the end ok you have a big file (let's say around 100mb) but at least your are sure you can run it (in case the model is not compatible anymore with the GAMA version).

**Comment by agrignard** [comment]
@hqnghi88 Does this feature is implemented? If it's included in GAMA 1.8 that would be great!

**Comment by agrignard** [comment]
This is how it is handle in processing, that would be great to have this feature to be able to produce release of models. they also propose to integrate java

<img width="381" alt="screen shot 2018-04-15 at 02 03 43" src="https://user-images.githubusercontent.com/3928502/38775471-668650ec-4051-11e8-910b-f9f92c5cfde3.png">

**Comment by sriramab** [comment]
Hallo, is development of this feature having any chance of resurrection ? 😞

**Comment by AlexisDrogoul** [comment]
Not to my knowledge, unless someone has started developing something ! :ami

**Comment by agrignard** [comment]
@hqnghi88 how difficult you think this feature can be? 

The ideal case would be to be able to generate a .app when you click on this .app it would be the equivalent of clicking on an experiment.

It's difficult for me to know what are the technical boundaries for that but it definetely can be a great feature

**Comment by hqnghi88** [comment]
Resurrection is here https://docs.google.com/document/d/1sbEYTiGn_cl8ObVVtJjinNDzu44Ycx3VgyKAObg7Dn0/edit?usp=sharing . Please check on macos, i have no way to review it except winOS.

**Comment by agrignard** [comment]
@hqnghi88 what about transfering this resurrection to a wiki page? I can then try to test it a bit

**Comment by agrignard** [comment]
Hi everyone,

I realive this old topic. I reread the hole discussion and I totally get the point of Alexis when he says:

https://github.com/gama-platform/gama/issues/1841#issuecomment-332863186

I get the point that the modeller should have access to the model and not only to the simulation. This is in a ideal research academic world. However there is also the final non academic user that might want to use the simulation and that can trust in it because a  relation has been made during the modelling process between the developper and him, in that case the final user might only wants to use the simulation as it has been designed and may not want to modify (or even take the risk to break it by modifying it) the model.  

@hqnghi88 I read the "resurection" proposition but if I understand well as soon as you open the test.gamr you arrive in the simulation perspective of the model but you can also switch to the model perspective right? than means we don't "hide" the model.

So I know it's a big philosophical discussion, to my mind I don't see the evil here , I understand the concern of hiding the model but in some case it could be a great way to diffuse our technology with some very basic simulations easy to run and to use. If we look at Processing for instance they are proposing this feature for a while https://github.com/processing/processing/wiki/Export-Info-and-Tips

For the technical detail of course I can also understand that it is not that easy to implement but having only the simulation perspective would be great!

@AlexisDrogoul , what do you think? Did your opinion change in one way or another on this topic?

**Comment by hqnghi88** [comment]
Dear Arnaud,
I am still on Alexis side, i dont want to "hide" regarding GPL or philosophical things. However i quit the discussion with the experimental plugin. So you test it, report it, request an enhancement. 
Yes in the current gamr , you can switch to modeling perspective and user can accidentally modify the model but it is easy to solve even with an locked simulation perspective.

**Comment by ptaillandier** [comment]
Hi,

I think the question of hiding or not the code depends largely on the
objective of the model. For example, for a pedagogical purpose, it can be
interesting in a first step to hide the code (for example in a
participatory context, where the participant should guess the dynamics of
some resources by observing the results of the simulation).

My point of view on this subject is that we should trust the modelers: it's
him/her who should take the responsibility to let the user access/modify or
not the code.

So for me, the best thing is to give to the modelers the possibility to
define an executable (here, just directly load a specific experiment of a
model), and for this executable, to define how the interface of GAMA will
look: possibility to access/edit the code or not, which menu, etc.

Cheers,

Patrick



Le dim. 3 févr. 2019 à 03:04, HUYNH Quang-Nghi <notifications@github.com> a
écrit :

**Comment by hqnghi88** [comment]
Hi Patrick,
Yes it would be nice if modelers take the responsibility. I was wonder only for the most feasible solution. I post another issue on defining how Gama will look.

**Comment by agrignard** [comment]
I agree with you but who else than the modeller can take the responsability to hide the code? This was my initial request, to give the modeller the opportunity to hide the code.

@hqnghi88 a first feasible solution would be only to have the simulation perspective active when the executable is launch

**Comment by hqnghi88** [comment]
Hello,
For me that feasible solution relates an enhance i posted and must be discussed there.
